bpf/kmesh/workload/include/authz.h

```diff
@@ -166,49 +255,144 @@ rule_match_check(Istio__Security__Rule *rule, struct xdp_info *info, struct bpf_
     return MATCHED;
 }
 
-static inline int
-do_auth(Istio__Security__Authorization *policy, struct xdp_info *info, struct bpf_sock_tuple *tuple_info)
+SEC("xdp_auth")
+int policy_check(struct xdp_md *ctx)
 {
-    void *rulesPtr = NULL;
-    Istio__Security__Rule *rule = NULL;
-    int matchFlag = 0;
-    __u32 i = 0;
+    struct match_context *match_ctx;
+    wl_policies_v *policies;
+    void *rulesPtr;
+    __u32 policyId;
+    Istio__Security__Authorization *policy;
+    struct bpf_sock_tuple tuple_key = {0};
+    struct xdp_info info = {0};
+    int ret;
+
+    if (construct_tuple_key(ctx, &tuple_key, &info) != PARSER_SUCC) {
+        BPF_LOG(ERR, AUTH, "policy_check, Failed to get tuple key");
+        return XDP_ABORTED;
+    }
 
-    if (policy->n_rules == 0) {
-        BPF_LOG(ERR, AUTH, "auth policy %s has no rules\n", kmesh_get_ptr_val(policy->name));
-        return AUTH_ALLOW;
+    match_ctx = bpf_map_lookup_elem(&kmesh_tc_info_map, &tuple_key);
+    if (!match_ctx) {
+        BPF_LOG(ERR, AUTH, "Failed to retrieve tailcall context from kmesh_tc_info_map");
+        return XDP_PASS;
     }
 
-    // Rules are OR-ed.
-    rulesPtr = kmesh_get_ptr_val(policy->rules);
-    if (!rulesPtr) {
-        BPF_LOG(ERR, AUTH, "failed to get rules from policy %s\n", kmesh_get_ptr_val(policy->name));
-        return AUTH_DENY;
+    policies = match_ctx->policies;
+    if (!policies) {
+        return XDP_PASS;
     }
 
+    // Safely access policyId and check if the policy exists
+    if (bpf_probe_read_kernel(&policyId, sizeof(policyId), (void *)(policies->policyIds + match_ctx->policy_index))
+        != 0) {
+        BPF_LOG(ERR, AUTH, "Failed to read policyId, throw it to user auth");
+        if (bpf_map_delete_elem(&kmesh_tc_info_map, &tuple_key) != 0) {
+            BPF_LOG(DEBUG, AUTH, "Failed to delete context from map");
+        }
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+    }
+    policy = map_lookup_authz(policyId);
+    if (!policy) {
+        // if no policy matches in xdp, thrown it to user auth
+        if (bpf_map_delete_elem(&kmesh_tc_info_map, &tuple_key) != 0) {
+            BPF_LOG(DEBUG, AUTH, "Failed to delete tailcall context from map");
+        }
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
```

Comment:
This code has multiple calls in the function, can it converge to one place?：
bpf_map_delete_elem(&kmesh_tc_info_map, &tuple_key)
bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);